src/mscorlib/src/System/Array.cs

```diff
@@ -2585,45 +2584,43 @@ sealed class SZArrayHelper {
         // This is a normal generic Enumerator for SZ arrays. It doesn't have any of the "this" stuff
         // that SZArrayHelper does.
         //
-        [Serializable] private sealed class SZGenericArrayEnumerator<T> : IEnumerator<T> {
+        [Serializable]
+        private sealed class SZGenericArrayEnumerator<T> : IEnumerator<T>
+        {
             private T[] _array;
             private int _index;
-            private int _endIndex; // cache array length, since it's a little slow.
 
-            // Passing -1 for endIndex so that MoveNext always returns false without mutating _index
-            internal static readonly SZGenericArrayEnumerator<T> Empty = new SZGenericArrayEnumerator<T>(null, -1);
+            internal static readonly SZGenericArrayEnumerator<T> Empty = new SZGenericArrayEnumerator<T>(Array.Empty<T>());
 
-            internal SZGenericArrayEnumerator(T[] array, int endIndex) {
-                // We allow passing null array in case of empty enumerator. 
-                Debug.Assert(array != null || endIndex == -1, "endIndex should be -1 in the case of a null array (for the empty enumerator).");
+            internal SZGenericArrayEnumerator(T[] array)
+            {
+                Debug.Assert(array != null);
                 _array = array;
                 _index = -1;
-                _endIndex = endIndex;
             }
     
-            public bool MoveNext() {
-                if (_index < _endIndex) {
-                    _index++;
-                    return (_index < _endIndex);
-                }
-                return false;
-            }
+            public bool MoveNext() => ++_index < _array.Length;
     
-            public T Current {
-                get {
-                    if (_index < 0) ThrowHelper.ThrowInvalidOperationException_InvalidOperation_EnumNotStarted();
-                    if (_index >= _endIndex) ThrowHelper.ThrowInvalidOperationException_InvalidOperation_EnumEnded();
+            public T Current
+            {
+                get
+                {
+                    if ((uint)_index >= (uint)_array.Length)
+                    {
+                        var exceptionResource = _index < 0 ?
+                            ExceptionResource.InvalidOperation_EnumNotStarted :
+                            ExceptionResource.InvalidOperation_EnumEnded;
+                        ThrowHelper.ThrowInvalidOperationException(exceptionResource);
+                    }
+
                     return _array[_index];
```

Comment:
As another micro-optimization, I wonder if it would be possible to use the Unsafe APIs to avoid a bounds check here? I just looked in Span, would `return Unsafe.Add(ref JitHelpers.GetArrayData(_array), _index)` be correct?